### PR TITLE
Hard fork localnet test

### PR DESCRIPTION
**Pull request by ejMina226**
Test for https://github.com/MinaProtocol/mina/issues/14709

Launches two local nodes and lets them run through all steps of hard-fork flow

### CODE REVIEW

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
TODO: move the test to nightlies

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by georgeee** [comment]
!ci-nightly-me

File: scripts/hardfork/test-helper.sh [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    if [[ "${slot_pe[$epoch]}" == "" ]] || [[ ${slot_pe[$epoch]} -lt ${f[$IX_SLOT]} ]]; then
+      slot_pe[$epoch]=${f[$IX_SLOT]}
+      snarked_hash_pe[$epoch]="${f[$IX_SNARKED_HASH]}"
+    fi
+  done
+  latest_str="${latest[*]}"
+  epochs="${!slot_pe[@]}"
+  epoch_str="${epochs[*]}"
+  snarked_hash_pe_str="${snarked_hash_pe[*]}"
+  echo "$max_slot,${epoch_str//${IFS:0:1}/:},${snarked_hash_pe_str//${IFS:0:1}/:},${latest_str//${IFS:0:1}/,}"
```
Comment (emberian) [inline scripts/hardfork/test-helper.sh:124]: shell programming... nothing like it

File: scripts/hardfork/test-helper.sh [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+  max_slot=0
+
+  # Read line by line, updating data above
+  while read l; do
+    IFS=, read -ra f <<< "$l"
+    if [[ $max_slot -lt ${f[$IX_SLOT]} ]]; then
+      max_slot=${f[$IX_SLOT]}
+    fi
+    if "${f[$IX_NON_EMPTY]}" && [[ ${latest[$IX_SLOT]} -lt ${f[$IX_SLOT]} ]]; then
```
Comment (emberian) [inline scripts/hardfork/test-helper.sh:111]: to my eye, this indexes f and runs it as as a command, surely the wrong thing?

File: scripts/hardfork/test-helper.sh (hunk 1 above)
Comment (georgeee) [reply scripts/hardfork/test-helper.sh:124]: That's true. I am not sure I did a good thing using Bash for the whole script, and at this point I started to doubt the choice :D

File: scripts/hardfork/test-helper.sh (hunk 2 above)
Comment (georgeee) [reply scripts/hardfork/test-helper.sh:111]: Yeah, and the value will be either `true` or `false` which are special-case commands

Comment (emberian) [reply scripts/hardfork/test-helper.sh:111]: 🙉  👍🏼

File: scripts/hardfork/test.sh [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+fi
+
+expected_fork_data="{\"fork\":{\"blockchain_length\":$latest_height,\"global_slot_since_genesis\":$latest_slot,\"state_hash\":\"$latest_shash\"},\"next_seed\":\"${latest_ne[$IX_NEXT_EPOCH_SEED]}\",\"staking_seed\":\"${latest_ne[$IX_CUR_EPOCH_SEED]}\"}"
+
+# 4. Check that no new blocks are created
+sleep 1m
+height1=$(get_height 10303)
+sleep 5m
+height2=$(get_height 10303)
+if [[ $(( height2 - height1 )) -gt 0 ]]; then
```
[the remark this answers is not in the document]
Comment (emberian) [reply scripts/hardfork/test.sh:110]: except there is no restarting! perfect.

**Review by emberian — APPROVED** [review]
scripts look good, don't forget to remove the tmp commits though.

**Comment by georgeee** [comment]
_Script is failing for SLOT_TX_END=96, investigating..._

**Comment by georgeee** [comment]
!ci-nightly-me

**Comment by emberian** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Comment by deepthiskumar** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!approved-for-mainnet

**Review by dkijania — APPROVED** [review]

**Pull request merged**